### ISSUE TITLE
Wrapping line with no spaces causes text "vibration"

### ISSUE THREAD

**Issue by Arcanemagus** [atom/atom#17268]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

If you have a single line of content that has _no_ spaces in it, have wrapping enabled, the line is long enough to be wrapped, and the file is _exactly_ the height of the window, the text will "vibrate" as the cursor is moved around it.

![2018-05-03_11-34-30](https://user-images.githubusercontent.com/427137/39596023-fe5c3c10-4ec5-11e8-8609-92b89f43647e.gif)

### Steps to Reproduce

1. Open Atom in Safe Mode.
2. Enable soft wrap of text
3. Write a line long enough to wrap
4. Fill the rest of the file with content (blank lines work) till the last line number is _right_ at the bottom of the window. There **must not** be a scroll bar, and there **must be** _less than_ ~50% of an empty line's worth of space between the last line number and the bottom. Shrinking the window till _just_ before it shows a scroll bar seems to be the easiest place.
    * Reproduces: ![image](https://user-images.githubusercontent.com/427137/39595765-39e6922c-4ec5-11e8-88cb-284044e7e155.png)
    * Fails to reproduce: ![image](https://user-images.githubusercontent.com/427137/39595797-52299398-4ec5-11e8-9ab6-1b430e7d8055.png)
5. Move the text cursor around, or try to select text with the mouse.

**Expected behavior:**

Text rendering to stay the same.

**Actual behavior:**

Text "vibrates" as the wrapping rapidly changes what should be wrapped or not.

**Reproduces how often:**

100% of the time.

### Versions

OS: Windows 10 x64
```
Atom    : 1.27.0-beta1
Electron: 1.7.11
Chrome  : 58.0.3029.110
Node    : 7.9.0
apm  1.19.0
npm  3.10.10
node 6.9.5 x64
atom 1.27.0-beta1
python 2.7.14
git 2.17.0.windows.1
visual studio 2015
```

### Additional Information

Possibly related to https://github.com/atom/atom/issues/13769.

**Comment by 50Wliu** [comment]
See also #15685, which only seems to occur when a vertical scrollbar is almost, but not quite, needed.

**Comment by Ben3eeE** [comment]
Did some debugging on this issue. I might be completely wrong because this code path is hard to follow but my current theory is below :sweat_smile: I think all three issues for this is the same issue.

---

It seems that we measure the horizontal scrollbar height and cache this measurement in `component.measurements`, This measurement is returned when `component.getHorizontalScrollbarHeight()` is called.

To calculate if we can scroll vertically we account for the horizontal scrollbar height using the same cached value. The problem is that if the horizontal scrollbar is not visible the height should be 0? If you calculate the real scrollbar height using `atom.workspace.getActiveTextEditor().component.refs.horizontalScrollbar.getRealScrollbarHeight()` it is 0.

So `canScrollVertically()` accounts for the horizontal scrollbar being visible which is wrong because it's not visible, this causes `canScrollVertically()` to return true because all conditions are met for being able to scroll vertically. This causes the update to change the soft wrap model column which changes the content width which causes the vertical scrollbar to be hidden again because one of the conditions are unmet which again changes the content width causing the flickering. This only happens if the content in the bottom of the editor is obscured by the not visible horizontal scrollbar.

The same problem exists for `canScrollHorizontally()` accounting for the vertical scrollbar width even when it's not visible.

So if we change `canScrollVertically()/canScrollHorizontally()` to use the real scrollbar height/width that changes with visibility the issue seems to be resolved. I don't think this is a proper solution and I don't know why the null check is needed. Seems `this.refs.verticalScrollbar` is not always present but it mostly is or why I used the other method when it's null. But it's something :sparkles: 

```diff
diff --git a/src/text-editor-component.js b/src/text-editor-component.js
index 2579bcfd0..94a61dc09 100644
--- a/src/text-editor-component.js
+++ b/src/text-editor-component.js
@@ -2653,9 +2653,10 @@ class TextEditorComponent {
     if (model.isMini()) return false
     if (model.getAutoHeight()) return false
     if (this.getContentHeight() > this.getScrollContainerHeight()) return true
+    const horizontalScrollbarHeight = this.refs.horizontalScrollbar ? this.refs.horizontalScrollbar.getRealScrollbarHeight() : this.getHorizontalScrollbarHeight()
     return (
       this.getContentWidth() > this.getScrollContainerWidth() &&
-      this.getContentHeight() > (this.getScrollContainerHeight() - this.getHorizontalScrollbarHeight())
+      this.getContentHeight() > (this.getScrollContainerHeight() - horizontalScrollbarHeight)
     )
   }

@@ -2665,9 +2666,10 @@ class TextEditorComponent {
     if (model.getAutoWidth()) return false
     if (model.isSoftWrapped()) return false
     if (this.getContentWidth() > this.getScrollContainerWidth()) return true
+    const verticalScrollbarWidth = this.refs.verticalScrollbar ? this.refs.verticalScrollbar.getRealScrollbarWidth() : this.getVerticalScrollbarWidth()
     return (
       this.getContentHeight() > this.getScrollContainerHeight() &&
-      this.getContentWidth() > (this.getScrollContainerWidth() - this.getVerticalScrollbarWidth())
+      this.getContentWidth() > (this.getScrollContainerWidth() - verticalScrollbarWidth)
     )
   }
```
